### PR TITLE
Add crypto_box::ChaChaBox support

### PR DESCRIPTION
**Pull request by alxiong**
_**status: ready for review**_

Currently, `crypto_box` uses `xsalsa20poly1035` by default and does not support customized aad field, see issue here: #224 

This PR add authenticated associated data (aad) support to `crypto_box` by **adding a separate `crypto_box::ChaChaBox` that wraps around `xchacha20poly1035`**, which is also what [libsodium](https://github.com/jedisct1/libsodium/tree/master/src/libsodium/crypto_box) supports.

- [x] pass `encrypt_in_place_detached` tests
- [x] add tests for non-empty aad field using `ChaChaBox`
- [x] add this option in `crypto_box` rustdoc

### CODE REVIEW

**Comment by alxiong** [comment]
@tarcieri may need your help here. I'm not so sure why tests `encrypt_in_place_detached` and `decrypt_in_place_detached` are failing right now.

I'll investigate myself, but I'm not so familiar with all the internal, so might take a while, wondering whether you have any idea~

also please feel free to give review for the 80%-baked code now, I'll iterate on your feedback.

thanks! :pray:

**Comment by alxiong** [comment]
I see where the problem is coming from:

```
---- xchacha20poly1305::encrypt_in_place_detached stdout ----
tag: [3, 216, 212, 59, 143, 166, 75, 41, 6, 42, 70, 153, 167, 173, 136, 82]
buffer: [224, 186, 214, 229, 38, 46, 233, 135, 39, 10, 14, 159, 120, 217, 16, 133, 30, 216, 156, 230, 110, 251, 70, 89, 16, 78, 76, 191, 116, 6, 45, 208, 130, 157, 84, 39, 59, 209, 78, 88, 151, 204, 164, 186, 152, 143, 226, 5, 245, 149, 240, 41, 52, 176, 197, 4, 116, 244, 24, 85, 218, 248, 243, 125, 160, 22, 20, 219, 73, 243, 137, 22, 135, 149, 177, 235, 103, 35, 33, 229, 14, 127, 186, 193, 236, 101, 199, 248, 34, 54, 45, 108, 167, 182, 205, 81, 75, 11, 61, 177, 129, 48, 111, 35, 125, 195, 68, 186, 18, 164, 27, 60, 101, 105, 233, 253, 157, 207, 206, 104, 8, 204, 244, 239, 106, 39, 194, 27, 124, 206, 42]
ciphertext: [224, 186, 214, 229, 38, 46, 233, 135, 39, 10, 14, 159, 120, 217, 16, 133, 30, 216, 156, 230, 110, 251, 70, 89, 16, 78, 76, 191, 116, 6, 45, 208, 130, 157, 84, 39, 59, 209, 78, 88, 151, 204, 164, 186, 152, 143, 226, 5, 245, 149, 240, 41, 52, 176, 197, 4, 116, 244, 24, 85, 218, 248, 243, 125, 160, 22, 20, 219, 73, 243, 137, 22, 135, 149, 177, 235, 103, 35, 33, 229, 14, 127, 186, 193, 236, 101, 199, 248, 34, 54, 45, 108, 167, 182, 205, 81, 75, 11, 61, 177, 129, 48, 111, 35, 125, 195, 68, 186, 18, 164, 27, 60, 101, 105, 233, 253, 157, 207, 206, 104, 8, 204, 244, 239, 106, 39, 194, 27, 124, 206, 42, 3, 216, 212, 59, 143, 166, 75, 41, 6, 42, 70, 153, 167, 173, 136, 82]
thread 'xchacha20poly1305::encrypt_in_place_detached' panicked at 'assertion failed: `(left == right)`
  left: `[224, 186, 214, 229, 38, 46, 233, 135, 39, 10, 14, 159, 120, 217, 16, 133]`,
 right: `[3, 216, 212, 59, 143, 166, 75, 41, 6, 42, 70, 153, 167, 173, 136, 82]`', crypto_box/tests/lib.rs:159:5
```
corresponds to test code:
```rust
            let mut buffer = $plaintext.to_vec();
            let tag = <$box>::new(&public_key, &secret_key)
                .encrypt_in_place_detached(nonce, b"", &mut buffer)
                .unwrap();
            println!("tag: {:?}", tag);
            println!("buffer: {:?}", buffer);
            println!("ciphertext: {:?}", $ciphertext);
            let (expected_tag, expected_ciphertext) = $ciphertext.split_at(16);
            assert_eq!(expected_tag, &tag[..]);
            assert_eq!(expected_ciphertext, &buffer[..]);
```

So, it seems that **when I use `salsa20`, the authenticated tag is _prepended_, whereas when I use `chacha20`, the tag is _appended_**!!

**Comment by tarcieri** [comment]
> So, it seems that when I use salsa20, the authenticated tag is prepended, whereas when I use chacha20, the tag is appended!!

Yep, XSalsa20Poly1305 uses a somewhat unusual prefix tag. (X)ChaCha20Poly1305 (the IETF version) uses a somewhat more standard postfix-tag.

> This discrepancy between salsa20poly1035 and xchacha20poly1035 should either be an accidental bug 🐞 or their respective RFC has diff regarding where to put the tag.

XSalsa20Poly1305 never had an RFC. Its spec was more or less the implementation in NaCl. Like you were experiencing with its lack of support for AAD, it's a little bit weird.

**Comment by tarcieri** [comment]
Cool, looks like you have it basically working. Will try to review soon.

File: crypto_box/src/lib.rs [hunk 1]
```diff
@@ -94,6 +94,58 @@
 //! assert_eq!(&plaintext[..], &decrypted_plaintext[..]);
 //! ```
 //!
+//! ## Customized Associated Data Usage
```
Comment (tarcieri) [inline crypto_box/src/lib.rs:97]: Maybe this section should be titled something like "Choosing XChaCha20Poly1305 vs XSalsa20Poly1305"

**Review by tarcieri — APPROVED** [review]
Looks good, thanks!

File: crypto_box/src/lib.rs [hunk 2]
```diff
@@ -94,7 +94,7 @@
 //! assert_eq!(&plaintext[..], &decrypted_plaintext[..]);
 //! ```
 //!
-//! ## Customized Associated Data Usage
+//! ## Choosing XChaCha20Poly1305 vs XSalsa20Poly1305
```
Comment (tarcieri) [inline crypto_box/src/lib.rs:97]: Thanks for changing this! I had a second thought though: it should probably be:

```suggestion
//! ## Choosing `ChaChaBox` vs `SalsaBox`
```

Sorry about that 😅

Comment (alxiong) [reply crypto_box/src/lib.rs:97]: haha right, you are right, that is clearer, no worry, updating now.

Comment (alxiong) [reply crypto_box/src/lib.rs:97]: do you want me to add "`SalsaBox(default)` ?

Comment (tarcieri) [reply crypto_box/src/lib.rs]: Nah, the `Box` type alias makes that clear, I think
